### PR TITLE
Reaver: Patch channel changing to nl80211 from wireless-extensions

### PR DESCRIPTION
**Pull request by damolp**
Maintainer:  @yousong

Compile tested / Run tested:
```
DISTRIB_ID='OpenWrt'
DISTRIB_RELEASE='23.05.0-rc3'
DISTRIB_REVISION='r23389-5deed175a5'
DISTRIB_TARGET='ramips/mt7621'
DISTRIB_ARCH='mipsel_24kc'
DISTRIB_DESCRIPTION='OpenWrt 23.05.0-rc3 r23389-5deed175a5'
DISTRIB_TAINTS='no-all'
```

Description:
Reaver still uses an embedded version of `libiw`, which relies on wireless-extensions which no longer works on `23.05.0` for channel changing
Upstream code resolved their issue in 2020 for this without fixing, hence this patch. 

upstream ref: https://github.com/t6x/reaver-wps-fork-t6x/issues/324

This patch makes reaver and wash channel changing work again on 5GHz and 2GHz.

---
Reaver 5GHz:
```
reaver -i wlan1 -b 00:11:22:33:44:55 -vv -5

Reaver v1.6.6 WiFi Protected Setup Attack Tool
Copyright (c) 2011, Tactical Network Solutions, Craig Heffner <cheffner@tacnetsol.com>

[+] Waiting for beacon from 00:11:22:33:44:55
[+] Switching wlan1 to channel 34
[+] Switching wlan1 to channel 36
[+] Switching wlan1 to channel 38
[+] Switching wlan1 to channel 40
[+] Received beacon from 00:11:22:33:44:55
[+] Vendor: Broadcom
[+] Trying pin "12345670"
[+] Sending authentication request
[+] Sending association request
[+] Associated with 00:11:22:33:44:55 (ESSID: OpenWRT)
^C
[+] Nothing done, nothing to save.
```

Wash 5GHz
```
wash -i wlan1 -5
BSSID               Ch  dBm  WPS  Lck  Vendor    ESSID
--------------------------------------------------------------------------------
00:AA:AA:AA:AA:AA   40  -56  2.0  Yes  Unknown   OpenWRT
00:BB:BB:BB:BB:BB   40  -87  2.0  Yes  AtherosC  Example
00:CC:CC:CC:CC:CC   40  -83  2.0  Yes  AtherosC  Example2
```

### CODE REVIEW

**Comment by neheb** [comment]
need PKG_RELEASE bump and patch refresh. make package/reaver/{clean,refresh}

File: net/reaver/patches/100-nl80211-channel-switching.patch
Comment (BKPepe) [inline net/reaver/patches/100-nl80211-channel-switching.patch:3]: > Upstream code resolved their issue in 2020 for this without fixing, hence this patch.

Why downstream should patch it? It does not make any sense. There isn't any Git header.
You are not author of this changes, so you should here add commit header, description, author, etc.

**Review by BKPepe — CHANGES_REQUESTED** [review]
Your commit says only something about patch, there are no details like in your OP. Nothing about adding new dependency, etc. Thus requesting changes.

File: net/reaver/Makefile
Comment (damolp) [reply net/reaver/Makefile:42]: Required for netlink APIs

File: net/reaver/Makefile
Comment (damolp) [reply net/reaver/Makefile:34]: Netlink APIs / Library. Upstream doesn't link against it hence required here.

The patch moves from wireless extension APIs to Netlink APIs.

I'll reword the commit description so that it is clearer.

File: net/reaver/patches/100-nl80211-channel-switching.patch
Comment (damolp) [reply net/reaver/patches/100-nl80211-channel-switching.patch:3]: Upstream won't fix, see https://github.com/t6x/reaver-wps-fork-t6x/issues/324

I just grabbed the code from one of my other tools that I have authored. I can add a header stating that I'm the author for the channel switch code and add a header for the aircrack-ng channel to frequency code.

Comment (BKPepe) [reply net/reaver/patches/100-nl80211-channel-switching.patch:3]: It's ``OpenWrt`` not ``OpenWRT``.

- Exactly as you do, I don't see a reason that we should merge this. We should rather remove this package once and for all. How many users have installed this package on their router (or rather, how many times it was downloaded)? Wireless extensions are disabled as you said in the kernel 5.15 and very few drivers use it.

From the issue which they won't fix:
> disabling WPS so reaver may already have become useless once the old wireless interface has been removed from the kernel.

We removed it (= disabled) from the kernel at least for OpenWrt. If you have your own usage, it can be compiled locally, isn't it?

- Upstream does not fix it, so why should we? Who is going to review it? Who will take care of issues, which it could bring?

- Of course, if you authored it, you should add it to the patch itself.

It is interesting that Ubuntu and as well Debian do not patch it.

This is up to @yousong to decide what we are going to do with this package as he imported it from oldpackages.

**Comment by yousong** [comment]
The patch should be accepted by upstream project first.  They know it more and it will benefit the community not just OpenWrt.

**Comment by BKPepe** [comment]
Putting this on hold.

**Comment by BKPepe** [comment]
Closing as it was said, the patch should be upstreamed or upstream developers should do something about it. Once it is handled there, you can reopen it or create a new pr.

**Comment by rofl0r** [comment]
a similar patch was just merged into upstream, adding support for both libnl3 and libnl-tiny. would be nice if someone could test it.

**Comment by radar27** [comment]
Thanks for the upstream merge @rofl0r. This was tested successfully with OpenWRT 23.05.5 (wireless extensions not present in kernel) on a Linksys EA8500 (ipq806x) and GL.iNet GL-AR300M16 (ath79) using [changes outlined](https://github.com/damolp/openwrt_packages/tree/19922cd7aded1faf8727d20847949b43818add79/net/reaver) with @damolp patches with a locally built version 1.6.6 PKG_RELEASE:=2.  The current package release (1.6.6-1) does not work on OpenWRT 23.05 nor 24.10.  Please consider a reopen/new pr to get this merged.

**Comment by rofl0r** [comment]
@radar27 did you test the changes from this PR or the one committed to upstream ? i was interested in getting the latter tested. that the current openwrt package release doesn't work (as in "channel switching doesn't work") is expected, unless they bump the git commit they use for it.
i won't open a new PR here - i expect that to be done by the folks involved in this PR.

**Pull request closed without merging**
